### PR TITLE
Cache workspace content

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Loading workspace content is slow and bound by rate limits.
This PR introduces a cache for workspace content.

### Linked issues
None

### Functionality
None

### Tests
- [x] added unit tests

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
let's have client code to keep one instance, that is passed down explicitly where it's needed:

```
cache = WorkspaceLruCache(ws, 2048)
foo_bar = cache.get_path('/foo/bar')
x = foo_bar / 'x'
_ = x.read_text()

x2 = cache.get_path('/foo/bar/x')
_ = x2.read_text()
```

File: src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+from collections import OrderedDict
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py]: done

File: src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        if isinstance(self._content, str):
+            return self
+        return StringIO(self._content.decode("utf-8"))
+
+
+# lru_cache won't let us invalidate cache entries
+# so we provide our own custom lru_cache
+class WorkspaceLruCache:
+
+    _WORKSPACE_CACHE_MAX_SIZE = 128
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py]: done

File: src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    # we need a wrapper to receive self in args
+    def wrapper(*args, **kwargs):
+        nonlocal cache
+        return cache(*args, **kwargs)
+
+    return wrapper
+
+
+class CachedWorkspacePath(WorkspacePath):
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py]: done

File: src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    def clear(cls):
+        cls._WORKSPACE_CACHE.clear()
+
+    @classmethod
+    def remove(cls, path: str):
+        if path in cls._WORKSPACE_CACHE:
+            cls._WORKSPACE_CACHE.pop(path)
+
+
+def workspace_lru_cache(func):
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py]: done

File: src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    def _as_string_io(self):
+        if isinstance(self._content, str):
+            return self
+        return StringIO(self._content.decode("utf-8"))
+
+
+# lru_cache won't let us invalidate cache entries
+# so we provide our own custom lru_cache
+class WorkspaceLruCache:
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py]: done

File: src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    _WORKSPACE_CACHE_MAX_SIZE = 128
+    _WORKSPACE_CACHE: OrderedDict[str, _CachedIO] = OrderedDict()
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/blueprint/CachedWorkspacePath.py]: done

File: tests/unit/blueprint/test_cached_workspace_path.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        _ = path.read_text()
+    assert ws.workspace.download.call_count == 1
+
+
+def test_download_is_called_again_after_unlink():
+    ws = mock_workspace_client()
+    ws.workspace.download.side_effect = lambda _, *, format: io.BytesIO("abc".encode())
+    path = CachedWorkspacePath(ws, "path")
+    _ = path.read_text()
+    path.unlink()
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply tests/unit/blueprint/test_cached_workspace_path.py]: done

File: tests/unit/conftest.py [hunk 8]
```diff
@@ -107,3 +109,9 @@ def mock_notebook_resolver():
     resolver = create_autospec(BaseNotebookResolver)
     resolver.resolve_notebook.return_value = None
     return resolver
+
+
+@pytest.fixture(autouse=True)
+def _clear_lru_cache():
+    WorkspaceLruCache.clear()
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply tests/unit/conftest.py]: Also, the above proposal (`cache.get_path(...)`) requires a change in blueprint (for `WorkspacePath.iterdir`)

[the remark this answers is not in the document]
Comment (nfx) [reply tests/unit/conftest.py]: > Also, the above proposal (cache.get_path(...)) requires a change in blueprint (for WorkspacePath.iterdir)

but you can inject cache and extend `iterdir` in a subclass

```
class CachedWorkspacePath:
  def __init__(self, cache, ws, path): ....
```
```

[the remark this answers is not in the document]
Comment (nfx) [reply tests/unit/conftest.py]: i'm a bit reluctant to the use of thread locals especially in python

File: src/databricks/labs/ucx/mixins/cached_workspace_path.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    def clear(cls):
+        cls._CACHE.clear()
+
+    @classmethod
+    def remove(cls, path: str):
+        if path in cls._CACHE:
+            cls._CACHE.pop(path)
+
+
+def _workspace_lru_cache(func):
```
Comment (nfx) [inline src/databricks/labs/ucx/mixins/cached_workspace_path.py:91]: do we need the decorator, if we can inject the cache as the property of `CachedPath`?

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/conftest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply tests/unit/conftest.py]: done

File: src/databricks/labs/ucx/mixins/cached_workspace_path.py (hunk 9 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/mixins/cached_workspace_path.py]: removed

**Review by nfx — APPROVED** [review]
nice!

**Pull request merged**